src/Tools/DynamoInstallDetective/ProductLookUp.cs

```diff
@@ -9,6 +9,118 @@
 
 namespace DynamoInstallDetective
 {
+    // Utility class for interacting with the windows registry.
+    internal static class RegUtils
+    {
+        // Utility class to enable/disable registry caching within a scope.
+        class RegistryCacher : IDisposable
+        {
+            public RegistryCacher()
+            {
+                EnableCache();
+            }
+
+            public void Dispose()
+            {
+                DisableCache();
+            }
+        }
+
+        public static string REG_KEY64 = @"SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\";
+        public static string REG_KEY32 = @"SOFTWARE\Wow6432Node\Microsoft\Windows\CurrentVersion\Uninstall\";
+
+        // Cached data from windows registry. Consists of a dictionary that maps The product name/code to Display Name (Tuple.item1) and Install Path(Tuple.item2)
+        // Dictionary<ProductCode, (DisplayName, InstallPath)>
+        private static Dictionary<string, Tuple<string, string>> cachedRecords;
+        private static bool cacheEnabled = false;
+
+        private static readonly object mutex = new object();
```

Comment:
maybe not worth investigating for the simple case, but have you looked at concurrentDictionary? 
https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2?view=net-5.0

Reply:
Totally forgot about those...thanks @mjkkirschner 
Will take a look and try to get rid of mutex

Reply:
Hmm..since I am allocating the dictionary and assigning null to it (and cacheEnabled boolean) , I do not think I can get rid of the lock altogether.
So it might not be very beneficial to introduce concurrentDictionary since that is meant to handle Add/Remove items in a concurrent scenario. The critical sections are only when we allocate the Dictionary, return a copy of it and  when we clear it up/make it null.

Reply:
thanks for considering and ruling it out!